### ISSUE TITLE
add reactor-parameterized tests for system events

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3198]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3198|
|Type|enhancement|
|Created|2008-04-20 17:44:58Z|
|Branch|https://github.com/twisted/twisted/tree/system-event-ordering-3198-2|

There should be tests for the reactor event system which are run against all reactors.

Attachments:

* [3198-iocp.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7f5/7f501ded67f7f6bd9df4c2f9731e25f5588f1ff9/0cd1a39876fb98343474fdf7ce6a512bb2bdc75c.diff) (1716 bytes) - added by therve on 2008-05-06 20:13:38Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3198 3198
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_system_event_ordering_3198_2 branches/system-event-ordering-3198-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1208713498000000 1208713498000000
changetime__1210950267000000 1210950267000000
version__None None
owner__ 
cc__therve cc__PenguinOfDoom
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [23342]) Refer to another ticket

refs [#3168](https://github.com/twisted/twisted/issues/3168)
refs [#3146](https://github.com/twisted/twisted/issues/3146)
refs [#3198](https://github.com/twisted/twisted/issues/3198)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23492]) Branching to 'system-event-ordering-3198-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

The branch may be in more or less reasonable shape.  However, iocpreactor fails at least one of the new tests such that the Windows slaves become unusable.  It'd be nice if someone with a Windows development environment could look into this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Attached patch fix run under windows with select reactor (when running the test of the iocp reactor). iocp wasn't passing the reactor argument properly, so the tests was using the global select reactor. I wonder if that can't happen somewhere without notice. Maybe we should delete twisted.internet.reactor during the tests?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [23529]) Apply's patch

Refs [#3198](https://github.com/twisted/twisted/issues/3198)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay, thanks to therve's iocp patch and a few more minor tweaks, I guess this is ready for review.

I didn't try to do anything as drastic as eliminate `twisted.internet.reactor` for the duration of the tests, since that would be hard.  It's certainly the case that there may be some shared state still hiding somewhere (in fact, it's not even really hiding - the process module is _hell_ of shared - it'd be great to do something about that, perhaps even in this branch... or perhaps in another one).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

 * the docstring of `IReactorCore.running` seems to miss a part
 * if I remove the `addSystemEventTrigger` in crash, no test in twisted.internet.test breaks. It seems vaguely useful but very indirectly
 * same thing for the _justStopped flag in stop, and in runUntilConcurrent
 * I don't see the use of the `_installSignalHandlers` flag: why not keep a system event? Is there a real reason

I suppose you realize the _started and _justStopped flags suck - just as the _stopped flag did. I **feel** it's a bit slightly tested, but well, it's not really a blocking point, because this branch already adds the ability to create good tests for reactors. So you can merge once the doc of the running flag is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

  * I'm not sure what you meant was missing from `IReactorCore.running`.
  * I added tests to `test_core.py` covering the `addSystemEventTrigger` in `crash
  * I deleted the `fireSystemEvent('shutdown')` in `crash`, since that was the point of `_justStopped`/`runUntilCurrent`.  Not sure how that call got there, I think I messed up the initial merge or something.  The manipulation of the `_justStopped` flag is now necessary for `test_core` to pass.
  * signal handlers could be installed with a system event, but then there'd be a second system event that needs to be installed in two places.  it could probably be done without too much mess, but I don't think it'll be much of a win.

`_started` and friends definitely suck.  I added a bit to the docstrings for these in hopes that we won't forget that we're unhappy with them. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [23634]) Merge system-event-ordering-3198-2

Author: exarkun
Reviewer: therve
Fixes: [#3198](https://github.com/twisted/twisted/issues/3198)

Add unit tests which are parameterized on the reactor implementation and
are can be run against any (or all) available reactor, regardless of the
reactor selected at trial invocation time.

Some tests are added for IReactorCore, IReactorTime, IReactorTCP, and
IReactorProcess.  None of these interfaces is completely covered by
the new tests; some old tests have been removed in favor of these new
ones.  Eventually most of the old tests should be converted to this
new style so that all reactors can be tested more easily.

Some small changes have been made to allow new reactors to be created
and used independently of the global reactor.  Since most reactor tests
are still written in the old style, this doesn't mean that multiple
reactors are now supported, only that they work well enough to pass the
small number of new tests which require this.

The "running" attribute is now documented (as most people treated it as
a public attribute already) and more thoroughly tested.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
